sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala

```diff
@@ -113,166 +114,112 @@ trait HashJoin {
+        Seq.empty
       }
     }
-    ret.iterator
   }
 
-  protected[this] def rightOuterIterator(
-      key: InternalRow,
-      leftIter: Iterable[InternalRow],
+  @transient private[this] lazy val nullRow = new GenericInternalRow(buildPlan.output.length)
+
+  protected[this] def outerIterator(
       joinedRow: JoinedRow,
-      resultProjection: InternalRow => InternalRow,
-      numOutputRows: LongSQLMetric): Iterator[InternalRow] = {
-    val ret: Iterable[InternalRow] = {
-      if (!key.anyNull) {
-        val temp = if (leftIter != null) {
-          leftIter.collect {
-            case l if boundCondition(joinedRow.withLeft(l)) => {
-              numOutputRows += 1
-              resultProjection(joinedRow).copy()
-            }
+      buildIter: Iterator[InternalRow]): Iterator[InternalRow] = {
+    new RowIterator {
+      private var found = false
+      override def advanceNext(): Boolean = {
+        while (buildIter != null && buildIter.hasNext) {
+          val nextBuildRow = buildIter.next()
+          if (boundCondition(joinedRow.withRight(nextBuildRow))) {
+            found = true
+            return true
           }
-        } else {
-          List.empty
         }
-        if (temp.isEmpty) {
-          numOutputRows += 1
-          resultProjection(joinedRow.withLeft(leftNullRow)) :: Nil
-        } else {
-          temp
+        if (!found) {
+          joinedRow.withRight(nullRow)
+          found = true
+          return true
         }
-      } else {
-        numOutputRows += 1
-        resultProjection(joinedRow.withLeft(leftNullRow)) :: Nil
+        false
       }
-    }
-    ret.iterator
+      override def getRow: InternalRow = joinedRow
+    }.toScala
   }
 
   protected def hashSemiJoin(
     streamIter: Iterator[InternalRow],
-    hashedRelation: HashedRelation,
-    numOutputRows: LongSQLMetric): Iterator[InternalRow] = {
+    hashedRelation: HashedRelation): Iterator[InternalRow] = {
     val joinKeys = streamSideKeyGenerator
     val joinedRow = new JoinedRow
     streamIter.filter { current =>
       val key = joinKeys(current)
-      lazy val rowBuffer = hashedRelation.get(key)
-      val r = !key.anyNull && rowBuffer != null && (condition.isEmpty || rowBuffer.exists {
+      lazy val buildIter = hashedRelation.get(key)
+      !key.anyNull && buildIter != null && (condition.isEmpty || buildIter.exists {
         (row: InternalRow) => boundCondition(joinedRow(current, row))
       })
-      if (r) numOutputRows += 1
-      r
+    }
+  }
+
+  protected def join(
+      streamedIter: Iterator[InternalRow],
+      hashed: HashedRelation,
+      numOutputRows: LongSQLMetric): Iterator[InternalRow] = {
+    val joinedRow = new JoinedRow()
+    val keyGenerator = streamSideKeyGenerator
+
+    val joinedIter = joinType match {
+      case Inner =>
+        innerJoin(streamedIter, hashed)
+
+      case LeftOuter | RightOuter =>
+        streamedIter.flatMap { currentRow =>
```

Comment:
hm i see why you named it outerIterator. can we move this and the outerIterator together just into a function called outerJoin?